feat(app): add history navigation with `[` and `]`, most recent command with `-`

tyzbit opened the pull request:
~~Go Back walks back through the history until at the oldest saved view~~

~~Last View switches between the current and previous views like how "cd -" works~~

EDIT: New description for the features

- Use square brackets `[` and `]` to navigate the command history back `[` and forward `]`.
- Use the minus/dash key `-` to switch between the current screen and the last one.

Example
While on the pod screen, type `:service` and enter. Hitting `[` then brings you back to the pod screen. Hitting `]` brings you forward in your history to the service screen. Type `:ingress` and hit enter (you are taken to the Ingresses screen), then hit `-`, you will go back to the service screen. Hit `-` again and you will go back to the Ingresses screen.

tyzbit commented:
Closes https://github.com/derailed/k9s/issues/1786 and https://github.com/derailed/k9s/issues/869

Features from these requests not included are:
- It is not possible to go forward again through the history. If you go back, the very last view you were on is removed from the history. `-` doesn't remove anything from history, just rearranges it.
- There is no overview to let you see the history and pick a specific view from it.

This PR implements the most powerful parts of the feature requests. Better navigation and richer history could be added later while keeping this functionality intact.

Supersedes https://github.com/derailed/k9s/pull/1090

KevinGimbel commented on internal/view/app.go:
```diff
@@ -691,6 +693,54 @@ func (a *App) helpCmd(evt *tcell.EventKey) *tcell.EventKey {
 
+// previousView returns to the view prior to the current one in the history
+func (a *App) previousView(evt *tcell.EventKey) *tcell.EventKey {
+	if evt != nil && evt.Rune() == 'b' && a.Prompt().InCmdMode() {
```
IMHO `evt.Rune() == 'b'` should use a variable for the comparison, not a hardcoded value. This would potentially allow rebinding the key in the future.

KevinGimbel commented on internal/view/app.go:
```diff
@@ -691,6 +693,54 @@ func (a *App) helpCmd(evt *tcell.EventKey) *tcell.EventKey {
+
+// lastView switches between the last view and this one a la `cd -`
+func (a *App) lastView(evt *tcell.EventKey) *tcell.EventKey {
+	if evt != nil && evt.Rune() == '-' && a.Prompt().InCmdMode() {
```
Same here as above.

KevinGimbel commented:
Tested the feature locally and I have some questions:

- When I look at containers (Select Pod -> Enter) I cannot go back to the pods with `-` or `b`
- Same when using Describe (`d`)

Is this intended? I assume these "views" are somehow different from the generic Object views (Pod, StatefulSet, Deployment, ...)?

In general I really like this feature! Especially `-` seems useful! Thank you @tyzbit!

tyzbit commented:
I would defer to someone more familiar with the codebase but `View`s seem to correlate mostly with Kubernetes resources (with the exception of `contexts`). Maybe the most accurate way to define them is the screens you get when you use a command (e.g. `:pods`). 

When you dive into a resource, you get a trail of breadcrumbs at the bottom.

<img width="912" alt="image" src="https://github.com/user-attachments/assets/9c61eeb9-5cbc-429e-bd36-568442a95d1e">

For demonstration, I went to `pods`, `namespaces` and then `deployments` and then selected a Deployment and hit Enter to go to the Pod, then hit enter again to go to the container, then I checked the `cmdHistory`:

<img width="287" alt="image" src="https://github.com/user-attachments/assets/c68eeb72-c137-475c-a1b8-50fb17d163db">

k9s doesn't add breadcrumb-related navigation to the command history at the present time.

In terms of addressing the difference in expectations, maybe I could update the documentation and help to clarify that Go Back and Last View only work with k8s resources and `contexts`?

KevinGimbel commented:
> In terms of addressing the difference in expectations, maybe I could update the documentation and help to clarify that Go Back and Last View only work with k8s resources and contexts?

I think that's a good  idea, just to make it clear what to expect from the feature :+1:

tyzbit commented:
I updated the README.

KevinGimbel commented:
Thanks! 

One tiny remark:

Your commit changed all `*` to `-` for lists - I assume some plugin or editor configuration on your side is making these changes? It makes reviewing the actual code changes hard, and I wonder if future changes will revert this due to different editor configs. 

My 2ct would be to revert these style-changes so only the new documentation is added, but I guess @derailed has the final call here :)

tyzbit commented:
@KevinGimbel those formatting changes were not intentional, I've changed my settings so it doesn't happen again. While I was at it, I squashed my commits.

KevinGimbel commented:
@tyzbit thank you! Looks good to me now. @derailed in my opinion this could be added to the next released :)

derailed commented on README.md:
```diff
@@ -355,6 +355,8 @@ K9s uses aliases to navigate most K8s resources.
 | To view and switch directly to another Kubernetes context (Last used view)      | `:`ctx context-name⏎          |                                                                        |
 | To view and switch to another Kubernetes namespace                              | `:`ns⏎                        |                                                                        |
+| To switch between the last active view/command (like how "cd -" works)          | `-`                           | Navigation that adds breadcrumbs to the bottom are not commands        |
+| To go back through previous views/commands                                      | `b`                           | Same as above                                                          |
```
I think 'b' is already in use in some views aka to run benchmarks ;(
Perhaps we can use '-' as the last 2 cmd toggle as you have here but then useuse '<' and '>' to nav the command history?

derailed commented on internal/view/app.go:
```diff
@@ -691,6 +693,54 @@ func (a *App) helpCmd(evt *tcell.EventKey) *tcell.EventKey {
 }
 
+// previousView returns to the view prior to the current one in the history
+func (a *App) previousView(evt *tcell.EventKey) *tcell.EventKey {
```
This should really be named previousCmd to avoid confusion

derailed commented on internal/view/app.go:
```diff
@@ -691,6 +693,54 @@ func (a *App) helpCmd(evt *tcell.EventKey) *tcell.EventKey {
+		return evt
+	}
+	cmds := a.cmdHistory.List()
+	if !(len(cmds) > 1) {
```
A dialog here feels a bit heavy, I think best to use a flash for this instance.
Also not keen on !(len(cmds) > 1.  This is len(cmds) <=1 return. 
But... as noted below better to defer these calls to cmdHistory.

derailed replied:
Additionally all these should be on cmdHistory apis. ie the logic for toggle last view and fwd/backwd should be methods on the history and not view logic imho i.ie toggleLastCmd, prevCmd, nextCmd, etc..

derailed replied:
Additionally if we surfaced a cmdHistory view all the necessary apis would be already there.

derailed commented on internal/view/app.go:
```diff
@@ -691,6 +693,54 @@ func (a *App) helpCmd(evt *tcell.EventKey) *tcell.EventKey {
+}
+
+// lastView switches between the last view and this one a la `cd -`
+func (a *App) lastView(evt *tcell.EventKey) *tcell.EventKey {
```
As above these are commands and not views since sub-views are not part of the cmd history.

derailed commented on internal/model/history.go:
```diff
@@ -23,14 +23,34 @@ func NewHistory(limit int) *History {
+	if len(n) > 1 {
+		// only one argument is expected
+		return false
+	} else if len(n) == 1 {
```
You are done here no need for the else clause.

derailed replied:
Also not keen on the vararg with only one option. Best to leave Pop as such and introduce PopN(n int)

derailed reviewed (changes requested):
@tyzbit Nice work! Good concept. Just a few items to clean up and simplify.

tyzbit commented:
Thank you for the great feedback. I'm working on addressing the comments; some planned changes:

- Navigate history with `[` and `]` instead of `B` (I prefer this over `<` and `>` because it is not necessary to hold shift)
- Add a view to see your history visually and pick a specific point in the history to go to. I think this will aid greatly in training users in using the new features.
- `cmdHistory` will keep a more accurate history of your commands such as saving the last command used as well as the current command (therefore duplicates are possible and the assumption that `history[0]` is the current command will be broken, requiring some other updates)

tyzbit commented:
I've implemented history navigation via `[` and `]` and switching to the last command using `-`. Right now, they're pretty similar because your last view will either be back one in the history or forward one, but I still plan to add a history page that lets you select a specific command out of your history - at that time, `lastCommand` will be much more useful. I opened https://github.com/derailed/k9s/issues/2895 to track that so as of this time, this PR is complete and all previous issues addressed.

merusso commented:
@derailed , any thoughts on this PR? Anything else needed to merge this

saiskee commented:
As someone that often demos at conferences using K9s, this would be a godsend!! @tyzbit Thank you so much for implementing this, and @derailed is there anything that is blocking this from going in?

mvillafuertem commented:
will it possible to change the key bindings to Vim bindings? (`Ctrl-o` to go back `Ctrl-i` to go forward)

tyzbit commented:
I don't think so, not any more than any other keybindings are customizable.

saiskee commented:
Really looking forward to this feature in the next release!

l-rossetti commented:
Hi, this feature would be very helpful. Any news on that? thanks

derailed reviewed (approved):
@tyzbit Nice! Thank you for the updates.

mloskot commented:
@tyzbit 
> Closes #1786 and #869

I'm wondering, shouldn't those two now be closed?

/cc @derailed

mvillafuertem commented:
Hi @tyzbit I’m having trouble with the new history navigation feature in K9s. When I’m in the Pods view and switch to Aliases, then try to navigate back using the history shortcut, it gets stuck on Aliases indefinitely. The navigation works fine elsewhere, but this issue happens consistently when switching to and from Aliases. Any idea why this might be happening?

<img width="634" alt="Screenshot 2025-02-26 at 15 22 48" src="https://github.com/user-attachments/assets/20ffa429-fc2d-4cd2-9950-4348433a0458" />

tyzbit commented:
I took a quick look at the code and the cause is here https://github.com/derailed/k9s/blob/e44c223e91c7fe5f6c77e3236272d1cf94bdd536/internal/view/command.go#L156-L158

Pod,Deployment,Service are not considered special commands but Aliases is (and others may be considered special as well). By breaking out here, that eventually causes another `aliases` to get pushed on the command stack, making the previous and current command `aliases` which makes it impossible to continue going back in the history.

I'll have more time to make a PR soon, but the fix shouldn't be too many lines of code if someone wants to take a stab at the issue.

tyzbit commented:
@mvillafuertem #3168 should fix your issue
